File: services/server/src/main/java/com/yaytsa/server/infrastructure/persistence/entity/TrackFeaturesEntity.java
package com.yaytsa.server.infrastructure.persistence.entity;

import jakarta.persistence.*;
import java.time.OffsetDateTime;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.Array;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "track_features")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class TrackFeaturesEntity {

  @Id
  @Column(name = "track_id")
  private UUID trackId;

  @OneToOne(fetch = FetchType.LAZY)
  @MapsId
  @JoinColumn(name = "track_id", foreignKey = @ForeignKey(name = "fk_track_features_item"))
  private ItemEntity item;

  @Column private Float bpm;

  @Column(name = "bpm_confidence")
  private Float bpmConfidence;

  @Column(name = "musical_key", length = 20)
  private String musicalKey;

  @Column(name = "key_confidence")
  private Float keyConfidence;

  @Column private Float energy;

  @Column(name = "loudness_integrated")
  private Float loudnessIntegrated;

  @Column(name = "loudness_range")
  private Float loudnessRange;

  @Column(name = "average_loudness")
  private Float averageLoudness;

  @Column private Float valence;

  @Column private Float arousal;

  @Column private Float danceability;

  @Column(name = "vocal_instrumental")
  private Float vocalInstrumental;

  @Column(name = "voice_gender", length = 10)
  private String voiceGender;

  @Column(name = "spectral_complexity")
  private Float spectralComplexity;

  @Column private Float dissonance;

  @Column(name = "onset_rate")
  private Float onsetRate;

  @Column(name = "intro_duration_sec")
  private Float introDurationSec;

  @JdbcTypeCode(SqlTypes.VECTOR)
  @Array(length = 1280)
  @Column(name = "embedding_discogs", columnDefinition = "vector(1280)")
  private float[] embeddingDiscogs;

  @JdbcTypeCode(SqlTypes.VECTOR)
  @Array(length = 200)
  @Column(name = "embedding_musicnn", columnDefinition = "vector(200)")
  private float[] embeddingMusicnn;

  @Column(name = "extracted_at")
  private OffsetDateTime extractedAt;

  @Column(name = "extractor_version", length = 20)
  private String extractorVersion;
}

File: infra-persistence/ml/src/main/kotlin/dev/yaytsa/persistence/ml/entity/TrackFeaturesEntity.kt
package dev.yaytsa.persistence.ml.entity

import dev.yaytsa.persistence.shared.converter.FloatArrayAttributeConverter
import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.Instant
import java.util.UUID

@Entity
@Table(name = "track_features", schema = "core_v2_ml")
class TrackFeaturesEntity(
    @Id
    @Column(name = "track_id")
    val trackId: UUID = UUID.randomUUID(),
    val bpm: Float? = null,
    @Column(name = "bpm_confidence")
    val bpmConfidence: Float? = null,
    @Column(name = "musical_key")
    val musicalKey: String? = null,
    @Column(name = "key_confidence")
    val keyConfidence: Float? = null,
    val energy: Float? = null,
    @Column(name = "loudness_integrated")
    val loudnessIntegrated: Float? = null,
    @Column(name = "loudness_range")
    val loudnessRange: Float? = null,
    @Column(name = "average_loudness")
    val averageLoudness: Float? = null,
    val valence: Float? = null,
    val arousal: Float? = null,
    val danceability: Float? = null,
    @Column(name = "vocal_instrumental")
    val vocalInstrumental: Float? = null,
    @Column(name = "voice_gender")
    val voiceGender: String? = null,
    @Column(name = "spectral_complexity")
    val spectralComplexity: Float? = null,
    val dissonance: Float? = null,
    @Column(name = "onset_rate")
    val onsetRate: Float? = null,
    @Column(name = "intro_duration_sec")
    val introDurationSec: Float? = null,
    @Convert(converter = FloatArrayAttributeConverter::class)
    @Column(name = "embedding_discogs", columnDefinition = "float[]")
    val embeddingDiscogs: FloatArray? = null,
    @Convert(converter = FloatArrayAttributeConverter::class)
    @Column(name = "embedding_musicnn", columnDefinition = "float[]")
    val embeddingMusicnn: FloatArray? = null,
    @Convert(converter = FloatArrayAttributeConverter::class)
    @Column(name = "embedding_clap", columnDefinition = "float[]")
    val embeddingClap: FloatArray? = null,
    @Convert(converter = FloatArrayAttributeConverter::class)
    @Column(name = "embedding_mert", columnDefinition = "float[]")
    val embeddingMert: FloatArray? = null,
    @Column(name = "extracted_at", nullable = false)
    val extractedAt: Instant = Instant.now(),
    @Column(name = "extractor_version", nullable = false)
    val extractorVersion: String = "",
)
